source/sigv4.c

```diff
@@ -794,6 +918,308 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+
+        buffRemaining = canonicalRequest->bufRemaining;
+        pCurrBufLoc = canonicalRequest->pBufCur;
+
+        for( index = 0; index < dataLen; index++ )
+        {
+            /* If the header field is not in canonical form already, we need to check
+             * whether this character represents a trimmable space. */
+            if( !( flags & SIGV4_HTTP_HEADERS_ARE_CANONICAL_FLAG ) &&
+                isTrimmableSpace( pData, index, dataLen, numOfBytesCopied ) )
+            {
+                /* Cannot copy trimmable space into canonical request buffer. */
+            }
+            /* Remaining buffer space should at least accommodate the character to copy and the trailing ":" */
+            else if( buffRemaining <= 1 )
+            {
+                status = SigV4InsufficientMemory;
+                break;
+            }
+            else
+            {
+                /* Lowercase header key only. '\n' character marks the ending of value. */
+                if( separator == '\n' )
+                {
+                    *pCurrBufLoc = ( pData[ index ] );
+                }
+                else
+                {
+                    *pCurrBufLoc = tolower( pData[ index ] );
+                }
+
+                pCurrBufLoc++;
+                numOfBytesCopied++;
+                buffRemaining -= 1;
+            }
+        }
+
+        /* Add the ending separating character passed to the function.
+         * Note: Space for the separator character is accounted for while copying
+         * header field data to canonical request buffer. */
+        if( status == SigV4Success )
+        {
+            assert( buffRemaining >= 1 );
+            *pCurrBufLoc = separator;
+            pCurrBufLoc++;
+            numOfBytesCopied++;
+            canonicalRequest->pBufCur = pCurrBufLoc;
+            canonicalRequest->bufRemaining = ( buffRemaining - 1 );
+        }
+
+        return status;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static SigV4Status_t appendSignedHeaders( size_t headerCount,
+                                              uint32_t flags,
+                                              CanonicalContext_t * canonicalRequest )
+    {
+        size_t headerIndex = 0, keyLen = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+        const char * headerKey;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+        assert( headerCount > 0 );
+
+        for( headerIndex = 0; headerIndex < headerCount; headerIndex++ )
+        {
+            assert( ( canonicalRequest->pHeadersLoc[ headerIndex ].key.pData ) != NULL );
+            keyLen = canonicalRequest->pHeadersLoc[ headerIndex ].key.dataLen;
+
+            headerKey = canonicalRequest->pHeadersLoc[ headerIndex ].key.pData;
+
+            sigV4Status = copyHeaderStringToCanonicalBuffer( headerKey, keyLen, flags, ';', canonicalRequest );
```

Comment:
Would be good to add a comment here that `;` is used to separate the "Signed Headers"